app/src/main/java/com/github/orkest/bluetooth/data/BluetoothServiceManager.kt

```diff
@@ -65,17 +80,100 @@ class BluetoothServiceManager(private var handler: Handler) : BluetoothInterface
+        serverConnection?.cancel()
+        serverConnection?.interrupt()
     }
 
 
+    @SuppressLint("MissingPermission")
+    private inner class AcceptThread : Thread() {
 
+        private val mmServerSocket: BluetoothServerSocket? by lazy(LazyThreadSafetyMode.NONE) {
+            bluetoothAdapter.listenUsingInsecureRfcommWithServiceRecord(NAME, MY_UUID)
+        }
 
+        override fun run() {
+            // Keep listening until exception occurs or a socket is returned.
+            var shouldLoop = true
+            while (shouldLoop) {
+                val socket: BluetoothSocket? = try {
+                    mmServerSocket?.accept()
+                } catch (e: IOException) {
+                    Log.e(TAG, "Socket's accept() method failed", e)
+                    shouldLoop = false
+                    null
+                }
+                socket?.also {
+                    //TODO: handle cancellation of this connection
+                    OrkestBluetoothCommunication(OrkestSocket(it), handler).apply {
+                        start()
+                        sendData(username)
+                    }
+                }
+            }
+        }
 
+        // Closes the connect socket and causes the thread to finish.
+        fun cancel() {
+            try {
+                mmServerSocket?.close()
+            } catch (e: IOException) {
+                Log.e(TAG, "Could not close the connect socket", e)
+            }
+        }
+    }
 
+    @SuppressLint("MissingPermission")
+    private inner class ConnectThread(device: BluetoothDevice) : Thread() {
+
+        private val mmSocket: BluetoothSocket? by lazy(LazyThreadSafetyMode.NONE) {
+            device.createRfcommSocketToServiceRecord(MY_UUID)
+        }
+
+        public override fun run() {
+            // Cancel discovery because it otherwise slows down the connection.
+            bluetoothAdapter.cancelDiscovery()
+
+            mmSocket?.let { socket ->
+                // Connect to the remote device through the socket. This call blocks
+                // until it succeeds or throws an exception.
+                socket.connect()
+
+                // The connection attempt succeeded.
+                // Create the connected thread to transfer data
+                //TODO: handle cancellation of this connection
+                val connection = OrkestBluetoothCommunication(OrkestSocket(socket),
+                                                                handler)
+                //Start the connected thread to receive the username of the other device
+                connection.start()
+                // Send the username of the current device to the other device
+                connection.sendData(username)
+            }
+        }
+
+        fun cancel() {
+            try {
+                mmSocket?.close()
+            } catch (e: IOException) {
+                Log.e(TAG, "Could not close the client socket", e)
+            }
+        }
+    }
 }
```

Comment:
You are handling interrupts and cancellations of threads but you should maybe also consider InterruptedException.

bluetoothAdapter should maybe be a val instead of a var

Reply:
bluetoothAdapter is late initialized so I can only have it as a var unfortunately